### ISSUE TITLE
Fatal Error: Uncaught Exception with message: [ssl:hostname:6380] in AbstractConnection.php

### ISSUE THREAD

**Issue by justin-frenzel** [nrk/predis#286]
I'm running my PHP app on an Azure Web App with Predis connecting to an Azure Redis Cache, under their SSL enabled 6380 port. I'm using the fork of this library that supports SSL.

My app consists of two parts, a client-side JS that sends AJAX requests to the PHP app for the HTML of the app (and in mosts cases, the App performs an API call before spitting out relevant HTML).

I'm using the Predis Session Handler, and also storing a variety of other non-user specific data. This is what my code looks like: -

	$redis = new Predis\Client($settings);

	try {
		$redis->connect();
	} catch (Predis\Connection\ConnectionException $exception) {
		new error("Redis.Connect()", array($redis, $exception));
		while (true) {
			usleep(200000);
			try {
				$redis->connect();
				break;
			}
			catch (Predis\Connection\ConnectionException $exception) {}
		}
	}

	$sessHandler = new Predis\Session\Handler($redis);
	session_set_save_handler($sessHandler);
	session_id($sessId);
	session_start();

In order to prevent the exception I'm getting from happening, I set up a while loop to run indefinitely to try and connect to Redis (assuming 99.99% availability of the Azure Redis Cache). If it fails, it will generate an entity in Azure Tables logging the fail.

It's generating tons of errors, it seems to fail to connect to Redis pretty much all the time, even though the Redis server itself has low load, with few clients connected. It always eventually connects, but with the while loop it ends up hemorrhaging performance.

Is there some way to address this issue? At other times, I get an error that that host/socket/port is in use and another connection can't be opened. Any ideas?

**Comment by justin-frenzel** [comment]
The more common error that I seem to be getting is this: -

Fatal error: Uncaught exception 'Predis\Connection\ConnectionException' with message 'Only one usage of each socket address (protocol/network address/port) is normally permitted.

How do I fix this?

**Comment by nrk** [comment]
@justin-frenzel the SSL feature is not battle tested and honestly I've never even tried using it on Windows systems yet. That said, on Windows the "_Only one usage of each socket address (protocol/network address/port) is normally permitted_" error usually suggests that the OS has no more available ports so it cannot reserve new ones for creating new connections (e.g. too many connections in `TIME_WAIT` state). See also https://msdn.microsoft.com/en-us/library/aa560610(v=bts.20).aspx for a description of the problem.

I'm not able to replicate the issue right now but if you are using PHP >= 5.4.0 you can modify the `Predis\Connection\StreamConnection` class by adding [here](https://github.com/nrk/predis/blob/d46de81d9139d7175dee50192c5e4033b6f95873/src/Connection/StreamConnection.php#L142) the following lines:

```php
$socket = socket_import_stream($resource);
socket_set_option($socket, SOL_TCP, SO_REUSEADDR, 1);
```

See if anything changes and please report back.

Thanks

**Comment by nrk** [comment]
I'm closing this issue due to the lack of feedback and the inability to reproduce it, but feel free to add more details if they can help tracing the source of the problem.

**Comment by MarioPerezEsteso** [comment]
We were having the same problem with the connections to Redis and we saw this issue. In spite of being closed, I would like to comment out that this can be solved in Laravel (afaik from 5.2 to 5.4 and predis 1.1) adding the option `'persistent' => true,` to the Redis connection in the file **config/database.php**:

```
'redis' => [
        'cluster' => false,
        'default' => [
            'host'     => env('CACHE_HOST', '127.0.0.1'),
            'port'     => env('CACHE_PORT', 6379),
            'password' => env('CACHE_PASSWORD', ''),
            'database' => 0,
            'persistent' => true, // This option
        ],
    ],
```
